### PR TITLE
feat(joins) Composite query pipeline Step 1: Make query pipeline stateless and composable 

### PR DESCRIPTION
**Pull request by fpacifici**
This is the plan to have a composite query pipeline:
1) This refactoring to adapt the pipeline structure
2) Introduce fork/join and split steps so that we can have a pipeline that splits into multiple paths
3) Populate the fork/join and splitter to run a composite query
4) Support selection of multiple query plans ranked 

This specifically introduces Query processing pipeline (which stops before executing the query). This will be useful for the composite query. 

Also I originally planned to introduce a pipeline segment abstraction to compose pipelines. Though I figured out that such abstraction did not add any value. Trying to build either a sequence generic pipeline, or a generic fork/join would add almost no value and make the pipeline more cumbersome. So I just removed it. It is still there and visible in the commits before https://github.com/getsentry/snuba/pull/1491/commits/0db1037ea98f886c2c2df37eaa48f23988c3fd9d

### CODE REVIEW

File: snuba/pipeline/query_pipeline.py [hunk 1]
```diff
@@ -1,44 +1,79 @@
...
+        Returns a pipeline to execute a query
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def build_processing_pipeline(self, request: Request) -> QueryProcessingPipeline:
+        """
+        Returns a pipeline that executes the processing phase of a query.
+
+        TODO: Split this method out of this class so that we can reuse
```
Comment (fpacifici) [inline snuba/pipeline/query_pipeline.py:75]: I am not 100% sure yet we will need the dataset to return the processing pipeline for the composite query. That will also depend on how we integrate the pipeline delegator (which will be decided tomorrow-ish). If the composite pipeline will be static, likely there will be no need for a pipeline builder at that level.

File: snuba/pipeline/__init__.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Generic, TypeVar
+
+TIn = TypeVar("TIn")
+TOut = TypeVar("TOut")
+
+
+class Segment(ABC, Generic[TIn, TOut]):
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/pipeline/__init__.py]: this does not exists anymore

File: snuba/pipeline/single_query_plan_pipeline.py [hunk 3]
```diff
@@ -1,58 +1,97 @@
...
+    This relies on a PreStrategyProcessingPipeline to perform the processing
+    and then relies on the execution strategy provided by the query plan to
+    actually execute the query.
+    """
+
+    def __init__(
+        self, runner: QueryRunner, query_plan_builder: ClickhouseQueryPlanBuilder,
+    ):
+        self.__runner = runner
+        self.__processing_pipeline = PreStrategyProcessingPipeline(query_plan_builder)
```
Comment (evanh) [inline snuba/pipeline/single_query_plan_pipeline.py:78]: It seems like an anti pattern to me to have a stage of the pipeline execute another pipeline. Having different pipelines that are all thin wrappers around other pipelines already confuses me without also having those pipelines take other pipelines as arguments and execute them.

File: snuba/pipeline/single_query_plan_pipeline.py [hunk 4]
```diff
@@ -1,58 +1,97 @@
...
 class SingleQueryPlanPipelineBuilder(QueryPipelineBuilder):
     def __init__(self, query_plan_builder: ClickhouseQueryPlanBuilder) -> None:
         self.__query_plan_builder = query_plan_builder
 
-    def build_pipeline(self, request: Request, runner: QueryRunner) -> QueryPipeline:
-        return SingleQueryPlanPipeline(request, runner, self.__query_plan_builder)
+    def build_execution_pipeline(
+        self, request: Request, runner: QueryRunner
+    ) -> QueryExecutionPipeline:
+        return SinglePlanExecutionPipeline(runner, self.__query_plan_builder)
```
Comment (evanh) [inline snuba/pipeline/single_query_plan_pipeline.py:94]: This is very confusing. This class has `build_execution_pipeline`, which builds a `SinglePlanExecutionPipeline` which in turn builds a `PreStrategyProcessingPipeline` which is a very thin wrapper around the `SinglePlanProcessingPipeline` and executes that as well as the buried execution strategy in the query plan. However it also has a `build_processing_pipeline` function that also returns a `PreStrategyProcessingPipeline`.

File: snuba/pipeline/single_query_plan_pipeline.py [hunk 5]
```diff
@@ -1,58 +1,97 @@
...
-    def execute(self) -> QueryResult:
-        # Execute entity processors
-        _execute_entity_processors(self.__request)
 
-        # Build and execute query plan
-        self.__query_plan = self.__query_plan_builder.build_plan(self.__request)
-        _execute_query_plan_processors(self.__query_plan, self.__request)
-        return self.__query_plan.execution_strategy.execute(
-            self.__query_plan.query, self.__request.settings, self.__runner
+class PreStrategyProcessingPipeline(SinglePlanProcessingPipeline):
```
Comment (evanh) [inline snuba/pipeline/single_query_plan_pipeline.py:49]: What is the purpose of this class? Why not call `SinglePlanProcessingPipeline(query_plan_builder, ClickhouseProcessorsExecutor(lambda plan: plan.plan_processors))` directly?

**Review by evanh — COMMENTED** [review]
I think this could be cleaned up to be more readable. Right now it's very hard to follow the flow through the pipelines.

File: snuba/pipeline/single_query_plan_pipeline.py (hunk 5 above)
Comment (fpacifici) [reply snuba/pipeline/single_query_plan_pipeline.py:49]: In the PR above this one I am adding a PostStrategyProcessingPipeline that executes all the query processors and skips the strategy. The issue is that the processing pipeline in a composite query is 75% the same as the single storage one, but then it skips the strategy and just runs the DB query processors. 
Unfortunately if I fit that part in this one as well, this PR will blow up in size.

File: snuba/pipeline/single_query_plan_pipeline.py (hunk 3 above)
Comment (fpacifici) [reply snuba/pipeline/single_query_plan_pipeline.py:78]: > It seems like an anti pattern to me to have a stage of the pipeline execute another pipeline.

I don't think I would agree with this. Why would it be an anti pattern? Pipelines can be composed and each component can have its own structure.
The issue is that when we throw composite queries and the delegator in the mix, our execution pipeline will have to do a lot of steps and can take different paths, many of them are similar. I think making them modular is inevitable. 

Also I don't think this is a thin wrapper around the processing pipeline. The execution pipeline adds a major step to the processing one, which is the execution of the query. The processing pipeline is its own module since it has to be used independently for the composite execution pipeline (which does not exist in this PR yet).

Would it help if I added trhe composite pipeline in this one as well (even though incomplete) ?

File: snuba/pipeline/single_query_plan_pipeline.py (hunk 4 above)
Comment (fpacifici) [reply snuba/pipeline/single_query_plan_pipeline.py:94]: Unfortunately the entity provides the query pipeline, in the execution case we need an SinglePlanExecutionPipeline from the pipeline builder, in the composite case we will need a QueryProcessingPipeline, thus two builder methods. We can use the builder for the execution pipeline as well, not sure how much more clear will be. Will try out.

File: snuba/pipeline/query_pipeline.py [hunk 6]
```diff
@@ -1,44 +1,79 @@
...
-
-    TODO: In future, the query pipeline wiil be composed of segments, which will
-    each perform individual processing steps and can be easily reused.
-    Examples of transformations performed by segments are entity processing, query
-    plan building and processing, storage processing, as well as splitters
-    and collectors to help break apart and reassemble joined queries.
     """
 
     @abstractmethod
-    def execute(self) -> QueryResult:
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/pipeline/query_pipeline.py]: Actually I ended up reverting this change, since basically all pipeline object are throw away objects that have to be built on the fly.
I think we should revisit this and actually make pipeline just functions, and pass around partially applied functions. There is no need to build an object.

File: snuba/pipeline/query_pipeline.py (hunk 1 above)
Comment (fpacifici) [reply snuba/pipeline/query_pipeline.py]: not relevant anymore

File: snuba/pipeline/single_query_plan_pipeline.py (hunk 4 above)
Comment (fpacifici) [reply snuba/pipeline/single_query_plan_pipeline.py]: This should not apply anymore since neither PreStrategyProcessingPipeline nor build_processing_pipeline exist anymore.

File: snuba/pipeline/single_query_plan_pipeline.py (hunk 5 above)
Comment (fpacifici) [reply snuba/pipeline/single_query_plan_pipeline.py]: Now we do not have a PreStrategyPipeline of any kind anymore, so this should not apply anymore

File: snuba/pipeline/single_query_plan_pipeline.py (hunk 3 above)
Comment (fpacifici) [reply snuba/pipeline/single_query_plan_pipeline.py]: Now this is quite different than it was, so probably the specific comment does not apply. Though we will need to compose pipelines from smaller parts anyway (the composite processing pipeline needs to run several entity processing pipelines (one per entity involved).

File: snuba/datasets/plans/query_plan.py [hunk 7]
```diff
@@ -33,7 +33,8 @@ class ClickhouseQueryPlan:
     """
 
     query: Query
-    plan_processors: Sequence[QueryProcessor]
+    plan_query_processors: Sequence[QueryProcessor]
+    db_query_processors: Sequence[QueryProcessor]
```
Comment (fpacifici) [inline snuba/datasets/plans/query_plan.py:37]: This did not exist before because there was never a case for executing the db query processor outside an execution strategy so db query processors were directly passed to the strategy. Now the split between these two is still required (when we run a query on an individual entity), but when we run a composite query, the query execution strategy is not provided by the storage anymore. At best it would be provided by a dataset, so we skip that step and we just need to run all processors on every subquery as if they were not split (in fact they are not, it is the main composite query that may be split).

File: snuba/pipeline/query_pipeline.py [hunk 8]
```diff
@@ -1,31 +1,41 @@
...
-    The QueryPipeline runs the query plan processors, executes the query plans and
-    returns the relevant result.
 
-    Most of the time, a single query plan is built by the SingleQueryPlanPipeline,
+    @abstractmethod
+    def execute(self) -> ClickhouseQueryPlan:
+        raise NotImplementedError
+
+
+class QueryExecutionPipeline(ABC):
```
Comment (fpacifici) [inline snuba/pipeline/query_pipeline.py:25]: @lynnagara  what if we turned all these classes into functions. The pipeline builder can return a callable (which would be a partially applied function).
In the end we always need to build these objects on the fly whether it is for the multiple storage pipeline or for the subqueries in the composite query. We never define a pipeline statically.

File: snuba/pipeline/query_pipeline.py [hunk 9]
```diff
@@ -35,10 +45,26 @@ def execute(self) -> QueryResult:
 
 class QueryPipelineBuilder(ABC):
```
Comment (fpacifici) [inline snuba/pipeline/query_pipeline.py:46]: I think this is unlikely to stay the way it is once we have the composite query in the picture.
There will most likely be an entity builder (that provides an execution and a processing pipeline for the single entity query and provided by an entity) and a dataset builder that provides an execution and a processing pipeline for requests. 
Though, how this finally look like will strongly depend on the pipeline delegator which is being built here: https://github.com/getsentry/snuba/pull/1465 So I will let that PR be finalized before making major changes to the builder.

File: snuba/pipeline/single_query_plan_pipeline.py [hunk 10]
```diff
@@ -35,24 +64,31 @@ def __init__(
...
-        return self.__query_plan.execution_strategy.execute(
-            self.__query_plan.query, self.__request.settings, self.__runner
-        )
+        settings = self.__request.settings
+        query = self.__request.query
+
+        execute_entity_processors(query, settings)
+
+        query_plan = self.__query_plan_builder.build_plan(query, settings)
+        execute_pre_strategy_processors(query_plan, settings)
```
Comment (lynnagara) [inline snuba/pipeline/single_query_plan_pipeline.py:73]: Does this mean we are executing the same query plan processors again? Isn't that already done by `execute_all_clickhouse_processors()` on line 47.

From what I gather, the processing pipeline works is supposed to run before the execution pipeline, though I'm not sure and the processing pipeline doesn't seem to be called anywhere yet. Is there a reason for that? I would have thought we'd still want to run that pipeline (a simplified version at least) for our simple query plans as well (not just relevant to joins)

Comment (fpacifici) [reply snuba/pipeline/single_query_plan_pipeline.py:73]: The SunglePlanExecutionPipeline does not execute the ProcessingPipeline. Originally I wanted to make the execution pipeline depend on the processing pipeline, but that turned out to be hard to grasp. See the comment from Evan (https://github.com/getsentry/snuba/pull/1491#discussion_r524384565). 
I did not double down on that route also because there really are three types of processing pipeline thus much less gain in reusing any of them as none of which can be reused.
1) A full processing pipeline. The one in this file. That is basically useful in tests and in the first iteration of composite queries. It executes all clickhouse query processing and skips the strategy.
2) A pre-strategy processing pipeline. Which would be used by the execution pipeline, because it only executes the processor before the strategy, instead of all of them.
3) A pre-plan processing which will be executed in the second iteration of the composite query, which allows building all the plans, merging back to select plans inn the same storage set and fork again.

So in the end we will have three use cases and three pipelines, then at that point, putting the pre strategy processing in its own pipeline when it is used by the execution only seems a lot less useful.

File: snuba/pipeline/single_query_plan_pipeline.py (hunk 4 above)
Comment (lynnagara) [reply snuba/pipeline/single_query_plan_pipeline.py]: I'm starting to think the entity shouldn't be the one providing the query pipeline anymore (in both the single entity and join cases). Could we just pass the entity/entities being requested to the pipeline builder and have it just build the appropriate pipeline. That way were are always building the pipeline in one place, we don't need the two separate methods to be called in different scenarios.

**Comment by lynnagara** [comment]
I think what took me the longest to wrap my head around here is that both our pipelines and query plans have now vastly different responsibilities depending on whether we are dealing with a query on a single or composite entity, and the specific responsibilities of each are now a bit fuzzy due to that.

In the single entity situation, the pipeline builds the query plan, executes the pre strategy processors, executes the query plan (which runs the db processors) and runs the query on ClickHouse. We get a result back out of the pipeline.

In the composite entity situation, the pipeline builds the query plan, executes all processors (which are now not part of the query plan) and just returns a query plan without doing the ClickHouse query. We only get a query plan out of the pipeline.

Obviously, there are different steps that need to go into each, but I wonder if its possible to align these 2 scenarios so that pipeline doesn't mean a completely different thing depending on the type of query we are dealing with.

**Comment by fpacifici** [comment]
> Obviously, there are different steps that need to go into each, but I wonder if its possible to align these 2 scenarios so that pipeline doesn't mean a completely different thing depending on the type of query we are dealing with.

Indeed, I would have preferred having a single pipeline as well. Though finally these are the two list of steps:
Simple:
1) entity processors
2) query plan building / translation
3) plan processors
4) execution strategy
4b) db processors
4c) execution

Composite
1) Join optimization 1
2) Push down (creates subqueries)
3) For each subquery:
3b) entity processors
3c) Plan selection and translation (produces multiple plans)
4) join
5) select valid plans combination
6) For each plan
6b) plan processors
6c) db processors
7) join
8) composite processors (when/if they will exist)
9) strategy (not the same as the simple one)
9b) join optimization 2
9c) execution

This means that:
- the simple execution pipeline is not a subset of the composite
- the simple processing pipeline is not a sub-component of the composite one (because of the re-join between plan and processor and because of the strategy)
- there are reusable components between the two (processor executor).

Trying to merge the simple one as a special case of the composite one could be possible, but it feels like all steps would have a big `if` at the beginning since the queries have different types and are managed differently, so that does not seem very useful as only 3 steps are reusable. We could build a pipeline template that defines the interface of each of the steps above and build two of them (simple and composite), but that breaks down because of the strategy. The processing part (which is the only reusable one) would have to be an individual box that we implement differently in the two cases.

So I am not very confident that would be of great value. 
What do you think ?

Also I think we should get rid of the classes here and just use Callable since pipelines are always built on the fly. What do you think about that ? https://github.com/getsentry/snuba/pull/1491#discussion_r524841055

File: snuba/pipeline/query_pipeline.py (hunk 8 above)
Comment (lynnagara) [reply snuba/pipeline/query_pipeline.py:25]: Turning this into a callable seems fine to me, it doesn't seem to do all that much except provide the execute method. Are you planning to keep an overall pipeline class though (ie. that represents the entire simple or composite pipeline), even if each of the segments of that pipeline are no longer pipelines themselves?

Comment (fpacifici) [reply snuba/pipeline/query_pipeline.py:25]: I don't think a pipeline class will be needed. A pipeline can easily be either:
`Callable[[Query, RequestSettings], QueryResult]`
or just
`Callable[[], QueryResult]`
depending on how we provide the function as it can be a partially applied one.

I came to this conclusion because we are using classes as partially applied functions:
- They are strictly throw away object as they have to be built on the fly and the state is not valid for a second call
- Using callables captures this stateless nature and reduces the confusion on when to initialize the pipeline objects.

**Comment by lynnagara** [comment]
I think the main issue isn't really that we don't have a single pipeline (though I think that would still be nice), it's that the way pipelines are used are completely different.

I think in both scenarios it would be good to aim to:
- build and execute both types of pipelines in the same way
- the two types of pipeline should return the same type (i.e. a QueryResult not a QueryPlan as SinglePlanQueryProcessingPipeline does)
- the differences in the two pipelines mostly encapsulated in the pipeline module
- if possible db_processors either always part of execution strategy or always not, not sometimes in and sometimes before

Do you think these are realistic?

Although, to go one step further on your outline of the steps above, the steps of the simple pipeline could be fairly similar to composite. To your point, what happens within each may end up being somewhat different though. On the surface at least this seems not any more different than all the different types of composite queries we will have to support at each step anyway.

1. Skip join optimization
2. Push down (not much to do, just create one subquery)
3. For that one subquery:
a) Entity processing
b) plan selection and translation
4. join (not much to do in the simple case)
5. select valid plan (there's just one)
6. For that one plan:
6a) Plan processing
6b) db processors
7. join (not much to do in the simple case)
8. Skip composite processors
9. execution strategy

Not sure if likely but if we wanted to do a join optimization at step 1 that turns the composite query into a simple one it becomes easy because they're the same and we don't need to switch over to the other type of pipeline.

**Comment by fpacifici** [comment]
Re:
- build and execute both types of pipelines in the same way
- the two types of pipeline should return the same type (i.e. a QueryResult not a QueryPlan as SinglePlanQueryProcessingPipeline does)

Actually that's the plan. SinglePlanQueryProcessingPipeline is NOT the composite execution pipeline. The composite execution pipeline will extend QueryExecutionPipeline and return QueryResult.
I wrapped the processing phase of a single entity query in a pipeline because that is a component of the composite pipeline (executed for each entity) and that needs to come from the same builder that generates the single plan execution pipeline.
The composite query execution pipeline will appear two PRs on top of this on the stack, but before that I need to be able to build a composite processing pipeline and inn order to do that I need a single plan processing pipeline that comes from the builder.

> if possible db_processors either always part of execution strategy or always not, not sometimes in and sometimes before

This though is impossible. 
In a simple query they are passed to the strategy. Not executed after the strategy, so that the strategy ca apply them to each of the split queries. Example: the column splitter needs to execute the prewhere processor also on the minimal query that is built on the fly. 
In the composite there may be a similar pattern if we introduce composite query processor and a composite strategy (no strong need for these now, maybe we will build them one day). But each individual subquery cannot go through the strategy and cannot be split, still we need to run all their processors (like the prewhere).

**Comment by fpacifici** [comment]
For context, this is where the processing pipeline introduced here is used (besides tests) https://github.com/getsentry/snuba/commit/18cab70fa7366b832fcbf7f4ba6c3d87bf9489e7#diff-4d6504b0a17d41fd1520c609c0e85a7b4d76b5b26b10d5782995543b875d986fR78-R84

**Comment by lynnagara** [comment]
So it sounds like db processing must stay in the execution strategy in the simple case due to the splitter. Are you implying that the execution strategy for composite queries will never be doing any splitting since the db processors in that case are run earlier in the pipeline?

**Comment by fpacifici** [comment]
This is a good point. Splitting a composite query is not something I would exclude, time splitting could make sense (and would not need to rerun subquery processors). Column splitting probably a lot less in its current concept. The algorithms for splitting would be radically different as they would have to deal with the main query and the subqueries. And none of the subquery could provide the logic to do the splitting in isolation.
I cannot think as of now to a good splitter that would need to rebuild a query from scratch and such that it would also recreate subqueries (and such that I am confident it would have a positive impact). 
By keeping the two list of processors separate (plan and db) we would still have the possibility to break the processing and execute the strategy before finalizing the subqueries (execute the db query processor). This is basically what we did in for the splitter already. It would not be a big change.
I would refrain from doing it until we have a reasonable use case for that as it would make the composite query processing quite a bit less intuitive and we still have a viable path to do that. Though open to do that if you disagree.

**Comment by fpacifici** [comment]
Different approach that should make the composite pipeline more similar to the simple pipeline:
0) The basics is to store all the details to run the query in the QueryPlan.
1) There is no queryProcessingPipeline anymore. There is a planner, produced by the pipeline builder.
2) The planner takes a logical query and produces a query plan (entity processors, query translation, etc. but it does not execute processors).
3) The planner is generic, so we can create a composite query plan that include multiple entity plans and the composite query, thus being able to execute the db processors in a composite query strategy. There it will be trivial.
4) The execution pipeline uses the planner to do the first part of the work and then executes processors and strategy.
5) In the next PR I will add the composite query plan and provide a pipeline_builder to the dataset that produces a planner and executor for that.

File: snuba/pipeline/simple_pipeline.py [hunk 11]
```diff
@@ -1,58 +1,85 @@
...
-        # Build and execute query plan
-        self.__query_plan = self.__query_plan_builder.build_plan(self.__request)
-        _execute_query_plan_processors(self.__query_plan, self.__request)
-        return self.__query_plan.execution_strategy.execute(
-            self.__query_plan.query, self.__request.settings, self.__runner
-        )
+        settings = self.__request.settings
+
+        query_plan = self.__query_planner.execute()
+        execute_pre_strategy_processors(query_plan, settings)
```
Comment (lynnagara) [inline snuba/pipeline/simple_pipeline.py:64]: Just want to confirm I understood this. This line won't be doing anything, right? Since the pre strategy (aka plan query) processors for a single entity query are always empty as per https://github.com/getsentry/snuba/blob/master/snuba/datasets/plans/single_storage.py#L138.
Pre strategy is a step that applies to composite queries only?

Comment (fpacifici) [reply snuba/pipeline/simple_pipeline.py:64]: the fact they are always empty is a coincidence. When we made the split we had use cases (we did not implement yet) for query processors that were supposed to run only once per plan. Maybe we can revisit that use case but as of now, we could have them.

File: snuba/pipeline/processors.py [hunk 12]
```diff
@@ -1,25 +1,74 @@
...
+    """
+    for clickhouse_processor in processors(query_plan):
         with sentry_sdk.start_span(
             description=type(clickhouse_processor).__name__, op="processor"
         ):
-            clickhouse_processor.process_query(query_plan.query, request.settings)
+            clickhouse_processor.process_query(query_plan.query, settings)
+
+
+def execute_pre_strategy_processors(
```
Comment (lynnagara) [inline snuba/pipeline/processors.py:32]: How about `execute_plan_processors` or so  for naming consistency?

**Review by lynnagara — APPROVED** [review]
Looks good to me

File: snuba/pipeline/processors.py (hunk 12 above)
Comment (fpacifici) [reply snuba/pipeline/processors.py:32]: ok.

File: snuba/pipeline/simple_pipeline.py (hunk 11 above)
Comment (fpacifici) [reply snuba/pipeline/simple_pipeline.py:64]: Oh now I remember. If we wanted to apply parameters that the splitter has to honor (like sampling), that has to happen per plan and not per db query.
